cache/cache.go

```diff
@@ -135,10 +203,19 @@ func (dc *directoryCache) Add(blobHash string, p []byte) {
 			fmt.Printf("Warning: could not create a cache file at %q: %v\n", c, err)
 			return
 		}
-		defer f.Close()
-		if n, err := f.Write(p); err != nil || n != len(p) {
-			fmt.Printf("Warning: failed to write cache: %d(wrote)/%d(expected): %v\n",
-				n, len(p), err)
+		want := b2.Len()
+		if _, err := io.CopyN(f, b2, int64(want)); err != nil {
```

Comment:
In order to minimize inconsistency, it should write to a temp file, then mv the temp file to the dest file.